# Data Exploration and Preprocessing

In [1]:
import pandas  as pd 
import numpy as np
data=pd.read_csv("C:\\Users\\HP\\OneDrive\\Desktop\\datascience\\Neural networks\\Neural networks\\Alphabets_data.csv")
data

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

Begin by loading and exploring the "Alphabets_data.csv" dataset. Summarize its key features such as the number of samples, features, and classes.

head () gives frist 5 rows of data

In [2]:
data.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [3]:
data.shape

(20000, 17)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


Execute necessary data preprocessing steps including data normalization, managing missing values.

checking missing values

In [5]:
data.isna().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

normalsing the data so separting the categriocal columns and numerical columns

In [6]:
x_df=data.iloc[:,1:17]
x_df

xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0         2     8      3       5      1     8    13      0      6      6   
1         5    12      3       7      2    10     5      5      4     13   
2         4    11      6       8      6    10     6      2      6     10   
3         7    11      6       6      3     5     9      4      6      4   
4         2     1      3       1      1     8     6      6      6      6   
...     ...   ...    ...     ...    ...   ...   ...    ...    ...    ...   
19995     2     2      3       3      2     7     7      7      6      6   
19996     7    10      8       8      4     4     8      6      9     12   
19997     6     9      6       7      5     6    11      3      7     11   
19998     2     3      4       2      1     8     7      2      6     10   
19999     4     9      6       6      2     9     5      3      1      8   

       x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          10       8      0       8      0       8  
1           3       9      2       8      4      10  
2           3       7      3       7      3       9  
3           4      10      6      10      2       8  
4           5       9      1       7      5      10  
...       ...     ...    ...     ...    ...     ...  
19995       6       4      2       8      3       7  
19996       9      13      2       9      3       7  
19997       9       5      2      12      2       4  
19998       6       8      1       9      5       8  
19999       1       8      2       7      2       8  

[20000 rows x 16 columns]

In [7]:
from sklearn.preprocessing import MinMaxScaler
scale=MinMaxScaler().fit_transform(x_df)
scale

array([[0.13333333, 0.53333333, 0.2       , ..., 0.53333333, 0.        ,
        0.53333333],
       [0.33333333, 0.8       , 0.2       , ..., 0.53333333, 0.26666667,
        0.66666667],
       [0.26666667, 0.73333333, 0.4       , ..., 0.46666667, 0.2       ,
        0.6       ],
       ...,
       [0.4       , 0.6       , 0.4       , ..., 0.8       , 0.13333333,
        0.26666667],
       [0.13333333, 0.2       , 0.26666667, ..., 0.6       , 0.33333333,
        0.53333333],
       [0.26666667, 0.6       , 0.4       , ..., 0.46666667, 0.13333333,
        0.53333333]])

encoding the categroical column

In [8]:
# encoding 
from sklearn.preprocessing import LabelEncoder
y_encode=LabelEncoder().fit_transform(data['letter'])
y_encode

array([19,  8,  3, ..., 19, 18,  0])

In [9]:
y_df=pd.DataFrame(y_encode,columns=['Letter'])
y_df

Letter
0          19
1           8
2           3
3          13
4           6
...       ...
19995       3
19996       2
19997      19
19998      18
19999       0

[20000 rows x 1 columns]

#  Model Implementation

dividing the data into train and test 

In [10]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_df,y_df,test_size=0.3,random_state=42)
print(f'x_train:{x_train}')
print('----'*10)
print(f'x_test:{x_test}')
print('----'*10)
print(f'y_train:{y_train}')
print('----'*10)
print(f'y_test:{y_test}')



x_train:       xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
17218     6    11      7       9      3     7     7     15      2      4   
15188     2     7      3       4      2     5    10      8      4      7   
11295     9    10      9       8      5     5     7      6      9      9   
19772     5     5      5       7      3     7     8     15      1      7   
13072     3     5      4       8      2     0    12      4      6     12   
...     ...   ...    ...     ...    ...   ...   ...    ...    ...    ...   
11284     5    11      4       6      3     7     7      4      3      9   
11964     3     7      5       5      4     9    11      6      4      6   
5390      3     4      6       3      2     9     6      2      8     10   
860       3     8      4       6      2     7     8      0      8     14   
15795     1     3      2       2      1    10     7      1      5     11   

       x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
17218       6       8    

A basic ANN with one hidden layer  

In [11]:
import tensorflow as tf
from tensorflow import keras
model1=keras.Sequential([
     keras.layers.Dense(64, input_shape=(x_train.shape[1],), activation='relu'),  # Hidden layer
    keras.layers.Dense(26, activation='softmax')  # Output layer
    ])
model1.compile(
    optimizer='adam',loss='sparse_categorical_crossentropy', metrics=['accuracy']
)
model1.fit(x_train,y_train,epochs=5)

Epoch 1/5
438/438 [==============================] - 3s 3ms/step - loss: 2.2743 - accuracy: 0.3884
Epoch 2/5
438/438 [==============================] - 1s 3ms/step - loss: 1.2791 - accuracy: 0.6649
Epoch 3/5
438/438 [==============================] - 1s 3ms/step - loss: 1.0653 - accuracy: 0.7196
Epoch 4/5
438/438 [==============================] - 2s 3ms/step - loss: 0.9687 - accuracy: 0.7426
Epoch 5/5
438/438 [==============================] - 1s 3ms/step - loss: 0.9000 - accuracy: 0.7563


Train your model on the training set and then use it to make predictions on the test set.

In [12]:
y_pred=model1.predict(x_test)
y_pred

188/188 [==============================] - 1s 2ms/step


array([[1.21233614e-04, 3.56220617e-03, 4.64195572e-03, ...,
        8.87227207e-02, 2.15685050e-05, 4.48393375e-01],
       [2.34367043e-01, 3.04676872e-03, 3.56974974e-02, ...,
        8.95902291e-02, 2.71322168e-02, 9.10821487e-04],
       [9.94269490e-01, 6.01454440e-06, 2.08069596e-05, ...,
        4.68816506e-05, 1.42407217e-07, 1.10067856e-06],
       ...,
       [1.05369970e-01, 3.70466097e-07, 4.76324378e-04, ...,
        7.40882751e-06, 3.93765482e-08, 1.69048162e-05],
       [1.99204963e-02, 7.01117739e-02, 3.69357876e-05, ...,
        3.31734726e-03, 6.26107285e-05, 1.13907486e-01],
       [1.09524734e-03, 2.86857896e-02, 7.41871263e-05, ...,
        1.46846510e-02, 4.94685839e-04, 9.62508318e-04]], dtype=float32)

# Hyperparameter Tuning

●	Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance.

In [14]:
import tensorflow as tf
from tensorflow import keras
model2=keras.Sequential([
    keras.layers.Dense(128 ,input_shape=(x_train.shape[1],),activation='relu'),#input layer
    keras.layers.Dense(64,activation='relu'),# hidden layer
    keras.layers.Dense(26,activation='softmax')# output layer
])

model2.compile(
    optimizer='adam',loss='sparse_categorical_crossentropy', metrics=['accuracy']
)
model2.fit(x_train,y_train,epochs=50)

Epoch 1/50
438/438 [==============================] - 3s 4ms/step - loss: 1.8334 - accuracy: 0.4871
Epoch 2/50
438/438 [==============================] - 2s 4ms/step - loss: 1.0152 - accuracy: 0.7125
Epoch 3/50
438/438 [==============================] - 2s 4ms/step - loss: 0.8303 - accuracy: 0.7601
Epoch 4/50
438/438 [==============================] - 2s 4ms/step - loss: 0.7236 - accuracy: 0.7882
Epoch 5/50
438/438 [==============================] - 2s 3ms/step - loss: 0.6389 - accuracy: 0.8100
Epoch 6/50
438/438 [==============================] - 2s 4ms/step - loss: 0.5793 - accuracy: 0.8279
Epoch 7/50
438/438 [==============================] - 2s 4ms/step - loss: 0.5184 - accuracy: 0.8446
Epoch 8/50
438/438 [==============================] - 2s 4ms/step - loss: 0.4830 - accuracy: 0.8538
Epoch 9/50
438/438 [==============================] - 2s 4ms/step - loss: 0.4466 - accuracy: 0.8656
Epoch 10/50
438/438 [==============================] - 2s 4ms/step - loss: 0.4210 - accuracy: 0.8698

In [15]:
pred=model2.predict(x_test)
pred

188/188 [==============================] - 1s 2ms/step


array([[1.07843216e-12, 2.21511975e-10, 1.31669419e-15, ...,
        4.19882208e-01, 4.63277541e-11, 5.51593781e-01],
       [1.17447591e-04, 3.11141477e-08, 7.75778688e-07, ...,
        6.55868575e-07, 3.57906149e-10, 6.32475165e-14],
       [1.00000000e+00, 2.28031378e-17, 5.85964370e-16, ...,
        3.09555957e-11, 2.26117050e-17, 1.10906623e-17],
       ...,
       [1.41382498e-13, 2.89242804e-28, 1.24326185e-08, ...,
        6.25582325e-19, 2.13616894e-22, 2.94260338e-17],
       [5.21151302e-03, 4.82389169e-06, 6.71773752e-24, ...,
        1.93791620e-05, 9.86770399e-09, 9.90290821e-01],
       [8.68772082e-14, 8.24158363e-12, 3.98157391e-14, ...,
        1.04065191e-06, 1.50489932e-09, 1.14790659e-16]], dtype=float32)

Adopt a structured approach like grid search or random search for hyperparameter tuning, documenting your methodology thoroughly.

i have used random search

In [17]:
from keras_tuner import RandomSearch
def build_model(hp):
    model = keras.Sequential([
        keras.layers.Dense(hp.Choice('units', [32, 64, 128]), activation=hp.Choice('activation', ['relu', 'tanh'])),
        keras.layers.Dense(26, activation='softmax')
    ])
    model.compile(optimizer=keras.optimizers.Adam(learning_rate=hp.Choice('learning_rate', [1e-2, 1e-3, 1e-4])),
                  loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model

# Configure and run the tuner
tuner = RandomSearch(build_model, objective='val_accuracy', max_trials=5, directory='.', project_name='alphabet_classification')
tuner.search(x_train, y_train, epochs=5, validation_split=0.2)

# Get and print the best hyperparameters
best_hps = tuner.get_best_hyperparameters(1)[0]
print(f"Best params: Units={best_hps.get('units')}, Activation={best_hps.get('activation')}, LR={best_hps.get('learning_rate')}")

# Train and evaluate the best model
best_model = tuner.hypermodel.build(best_hps)
best_model.fit(x_train, y_train, epochs=10, validation_split=0.2)
print(f"Test Accuracy: {best_model.evaluate(x_test, y_test)[1]:.4f}")


Reloading Tuner from .\alphabet_classification\tuner0.json
Best params: Units=128, Activation=relu, LR=0.01
Epoch 1/10
350/350 [==============================] - 3s 5ms/step - loss: 1.4631 - accuracy: 0.5736 - val_loss: 1.0279 - val_accuracy: 0.6921
Epoch 2/10
350/350 [==============================] - 1s 4ms/step - loss: 0.9166 - accuracy: 0.7190 - val_loss: 0.8148 - val_accuracy: 0.7579
Epoch 3/10
350/350 [==============================] - 2s 5ms/step - loss: 0.8038 - accuracy: 0.7495 - val_loss: 0.7802 - val_accuracy: 0.7611
Epoch 4/10
350/350 [==============================] - 2s 4ms/step - loss: 0.7352 - accuracy: 0.7676 - val_loss: 0.9013 - val_accuracy: 0.7168
Epoch 5/10
350/350 [==============================] - 1s 4ms/step - loss: 0.6594 - accuracy: 0.7981 - val_loss: 0.6072 - val_accuracy: 0.8179
Epoch 6/10
350/350 [==============================] - 1s 4ms/step - loss: 0.6085 - accuracy: 0.8051 - val_loss: 0.6313 - val_accuracy: 0.7993
Epoch 7/10
350/350 [====================

In [21]:
y_pred2 = best_model.predict(x_test)
y_pred2

188/188 [==============================] - 1s 2ms/step


array([[3.0193176e-10, 1.1088617e-05, 5.1985324e-08, ..., 7.8009389e-02,
        4.1333164e-09, 8.4823757e-01],
       [2.5069666e-01, 7.7924551e-03, 4.9415877e-04, ..., 1.4563702e-01,
        2.6678541e-04, 4.1249525e-03],
       [1.0000000e+00, 1.5058721e-19, 2.2500713e-24, ..., 9.4397545e-12,
        8.6783615e-20, 4.7992997e-15],
       ...,
       [5.1381083e-14, 0.0000000e+00, 8.9158523e-17, ..., 3.8138407e-18,
        8.3753082e-34, 3.4982148e-26],
       [8.6258503e-04, 1.2545756e-06, 1.2688123e-19, ..., 1.2881665e-06,
        7.0875380e-08, 9.7748780e-01],
       [1.4893059e-06, 1.1541476e-07, 2.8173051e-09, ..., 2.8731747e-04,
        5.3514608e-08, 4.9869204e-08]], dtype=float32)

# Evaluation
Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.


performance using one hidden layer 

In [19]:
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

# Convert the predictions to class labels (argmax for classification)
y_pred_classes = np.argmax(y_pred, axis=1)

# Compute accuracy, precision, recall, and F1-score
accuracy = accuracy_score(y_test, y_pred_classes)
precision = precision_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class
recall = recall_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class
f1 = f1_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class

# Print the results
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Accuracy: 0.7592
Precision: 0.7715
Recall: 0.7592
F1-Score: 0.7567


performance after using input layer hidden layer increasing the epochs values and using activations relu softmax 

In [20]:
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

# Convert the predictions to class labels (argmax for classification)
y_pred_classes = np.argmax(pred, axis=1)

# Compute accuracy, precision, recall, and F1-score
accuracy = accuracy_score(y_test, y_pred_classes)
precision = precision_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class
recall = recall_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class
f1 = f1_score(y_test, y_pred_classes, average='weighted')  # 'weighted' for multi-class

# Print the results
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Accuracy: 0.9383
Precision: 0.9402
Recall: 0.9383
F1-Score: 0.9383


performance after using hyperpaarnter tunning random search

In [22]:
y_pred_classes = tf.argmax(y_pred2, axis=1)  # Convert probabilities to class labels

# Calculate precision, recall, and F1-score
precision = precision_score(y_test, y_pred_classes, average='weighted')
recall = recall_score(y_test, y_pred_classes, average='weighted')
f1 = f1_score(y_test, y_pred_classes, average='weighted')

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Precision: 0.8383
Recall: 0.8240
F1-Score: 0.8241


Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.


Performance Comparison:

Model with One Hidden Layer (Default Hyperparameters):

Accuracy: 0.7592

Precision: 0.7715

Recall: 0.7592

F1-Score: 0.7567

Analysis: The default model performs decently but has room for improvement. Training with more epochs and optimizing activation functions could help.

Model with Increased Epochs and ReLU Activation:

Accuracy: 0.9383

Precision: 0.9402
Recall: 0.9383

F1-Score: 0.9383

Analysis: Increasing epochs and using ReLU for the hidden layer and Softmax for output improved performance significantly, showing the benefits of more training and better activation functions.

Model After Hyperparameter Tuning (Random Search):

Precision: 0.8383

Recall: 0.8240

F1-Score: 0.8241

Analysis: Hyperparameter tuning improved the model, but the performance was slightly lower than the manually optimized model. This suggests that more
trials or a finer search might yield better results.

Conclusion:

Best Performance: Achieved by increasing epochs and using ReLU and Softmax activations.

Hyperparameter Tuning: Random Search provided moderate improvement, but manually adjusting training settings had a bigger impact. Hyperparameter tuning can be powerful but may require more thorough exploration.